source/math/FloatConvexHull2D.ooc

```diff
@@ -33,42 +35,99 @@ FloatConvexHull2D: class {
 		this _points add(box rightBottom)
 		this _points add(box rightTop)
 	}
+	free: override func {
+		this _points free()
+		super()
+	}
 	contains: func ~Point (point: FloatPoint2D) -> Bool {
 		result := true
-		for (i in 0 .. this count - 1)
-			if (This _isOnLeft(this _points[i], this _points[i + 1], point)) {
+		for (i in 0 .. this count)
+			if (result && This _isOnLeft(this _points[i], this _points[(i + 1) % this count], point))
 				result = false
-				break
-			}
-		if (This _isOnLeft(this _points[this count - 1], this _points[0], point))
-			result = false
 		result
 	}
 	contains: func ~ConvexHull (other: This) -> Bool {
 		result := true
 		for (i in 0 .. other count)
-			if (!this contains(other points[i])) {
+			if (result && !this contains(other _points[i]))
 				result = false
-				break
-			}
 		result
 	}
 	contains: func ~FloatBox2D (box: FloatBox2D) -> Bool {
 		this contains(box leftBottom) && this contains(box rightBottom) && this contains(box leftTop) && this contains(box rightTop)
 	}
 	transform: func (transform: FloatTransform2D) -> This {
-		// TODO: Re-compute convex hull from new points if necessary, i.e. if transform changes order of points
 		newPoints := VectorList<FloatPoint2D> new(this count)
 		for (i in 0 .. this count)
 			newPoints add(transform * this _points[i])
 		This new(newPoints)
 	}
-	_isOnLeft: static func (leftPoint, rightPoint, queryPoint: FloatPoint2D) -> Bool {
-		(rightPoint x - leftPoint x) * (queryPoint y - leftPoint y) > (rightPoint y - leftPoint y) * (queryPoint x - leftPoint x)
+	computeHull: func {
+		if (this count > 2) {
+			// Uses the Quickhull algorithm if more than two points, average complexity O(n log n)
+			// http://www.cse.yorku.ca/~aaw/Hang/quick_points/Algorithm.html
+			points := this _points copy()
+			this _points clear()
+			leftMostIndex := 0
+			rightMostIndex := 0
+			for (i in 0 .. points count)
+				if (points[i] x < points[leftMostIndex] x)
+					leftMostIndex = i
+				else if (points[i] x > points[rightMostIndex] x)
+					rightMostIndex = i
+			leftEndpoint := points[leftMostIndex]
+			rightEndpoint := points[rightMostIndex]
+			leftSet := VectorList<FloatPoint2D> new()
+			rightSet := VectorList<FloatPoint2D> new()
+			for (i in 0 .. points count)
+				if (i != leftMostIndex && i != rightMostIndex)
+					if (This _isOnLeft(points[leftMostIndex], points[rightMostIndex], points[i]))
+						leftSet add(points[i])
+					else
+						rightSet add(points[i])
+			this _points add(leftEndpoint)
+			this _findHull(leftSet, leftEndpoint, rightEndpoint)
+			this _points add(rightEndpoint)
+			this _findHull(rightSet, rightEndpoint, leftEndpoint)
+			points free()
+			leftSet free()
+			rightSet free()
+		}
 	}
-	free: override func {
-		this _points free()
-		super()
+	_findHull: func (currentSet: VectorList<FloatPoint2D>, leftPoint, rightPoint: FloatPoint2D) {
```

Comment:
This name is a bit confusing, it sounds like it would return a hull or something but it actually adds points to the current hull. I don't think this is a big deal though.